### PR TITLE
Add secure connection and relay sub-protocol

### PR DESCRIPTION
**Pull request by ghost**
1. Add secure connection
   Implements handshake protocol based on [Noise Protocol Frame](https://noiseprotocol.org/noise.html), and we use the XX pattern. With this handshake, a secure connection is build from an insecure connection which implements the `net.Conn` interface.
2. Relay sub-protocol
   This PR defines the message types of the Relay sub-protocol.

### CODE REVIEW

File: secure/crypto.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package secure
+
+import (
+	"errors"
+)
+
+func (s *SecureSession) encrypt(out, plaintext []byte) ([]byte, error) {
+	if s.enc == nil {
+		return nil, errors.New("cannot encrypt, handshake incomplete")
```
Comment (adahanaresh) [inline secure/crypto.go]: please replace the error output with a more clear error root cause

File: secure/crypto.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	}
+	return s.enc.Encrypt(out, nil, plaintext), nil
+}
+
+func (s *SecureSession) decrypt(out, ciphertext []byte) ([]byte, error) {
```
Comment (adahanaresh) [inline secure/crypto.go:14]: same as encrypto

**Review by adahanaresh — CHANGES_REQUESTED** [review]
see comment in code

File: secure/crypto.go (hunk 2 above)
Comment (ghost) [reply secure/crypto.go:14]: No handshake or handshake failure at any phase may cause this error. Cause of handshake failure is returned when using `NewSecureSession` to instantiate a `SecureSession`.

Same as above.

Comment (ghost) [reply secure/crypto.go:14]: error output changed

**Review by adahanaresh — APPROVED** [review]

**Comment by adahanaresh** [comment]
please update code that need to be changed at first before your new commit

**Pull request merged**